### PR TITLE
[PATCH API-NEXT v4] api: packet, cls, vxlan: vxlan and inner header parsing support

### PR DESCRIPTION
**Pull request by bala-manoharan**
1. Enhances packet parsing engine to handle inner header parsing for VXLAN packets
2. Implements classification PMR verification for PMR_LD_VNI
3. Adds test case for PMR LD_VNI

### CODE REVIEW

**Comment by Bill-Fischofer-Linaro** [comment]
make distcheck test is failing:
```
  CC       odp_classification.lo
In file included from ../../../platform/linux-generic/odp_classification.c:19:0:
../../../platform/linux-generic/include/odp_classification_inlines.h:27:29: fatal error: protocols/vxlan.h: No such file or directory
 #include <protocols/vxlan.h>
                             ^
compilation terminated.
make[2]: *** [odp_classification.lo] Error 1
make[2]: Leaving directory `/home/travis/build/Linaro/odp/opendataplane-1.16.0.0/_build/platform/linux-generic'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/travis/build/Linaro/odp/opendataplane-1.16.0.0/_build'
make: *** [distcheck] Error 1
The command "sudo PATH="$PATH" LD_LIBRARY_PATH="$HOME/cunit-install/$CROSS_ARCH/lib:$LD_LIBRARY_PATH" make distcheck" exited with 2.
```
Have to make sure newly added files are included in the appropriate `Makefile.am` file lists.

File: platform/linux-generic/odp_packet.c [hunk 1]
```diff
@@ -1996,7 +2016,10 @@ static inline void parse_udp(packet_parser_t *prs,
 
 	if (offset)
 		*offset   += sizeof(_odp_udphdr_t);
-	*parseptr += sizeof(_odp_udphdr_t);
+	*parseptr += _ODP_UDPHDR_LEN;
+	dport = odp_be_to_cpu_16(udp->dst_port);
+	if (dport == _ODP_UDP_VXLAN_PORT)
+		parse_vxlan(prs, in_p, parseptr, offset, frame_len, seg_len);
```
Comment (Bill-Fischofer-Linaro) [inline platform/linux-generic/odp_packet.c]: Need to validate that the packet (and specified UDP length) is large enough to contain the VXLAN header and the encapsulated packet.

File: test/validation/api/classification/odp_classification_common.c [hunk 2]
```diff
@@ -107,8 +107,23 @@ int cls_pkt_set_seq(odp_packet_t pkt)
 	CU_ASSERT_FATAL(offset != ODP_PACKET_OFFSET_INVALID);
 
 	if (ip->proto == ODPH_IPPROTO_UDP)
-		status = odp_packet_copy_from_mem(pkt, offset + ODPH_UDPHDR_LEN,
-						  sizeof(data), &data);
+		if (odp_packet_has_vxlan(pkt)) {
+			/* TODO: Inner header parsing is not supported
+			 * in this ODP release. Hence the packet parsing
+			 * is skipped. This code has to be modifies once
```
Comment (Bill-Fischofer-Linaro) [inline test/validation/api/classification/odp_classification_common.c]: Typo: modified

Comment (Bill-Fischofer-Linaro) [reply test/validation/api/classification/odp_classification_common.c]: The "TODO" can't stand by itself. Sounds like we need an additional `odp_classification_capability()` bit to indicate whether inner packets are included in the parse?

File: helper/include/odp/helper/vxlan.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+#define ODPH_UDP_VXLAN_PORT 4789
+
+/** VNI to VXLAN header format */
+#define ODPH_VNI_VXLAN(vxlan_vni) (vxlan_vni << 8)
+
+/** VXLAN header */
+typedef struct ODP_PACKED {
+	uint8_t flags; /**< Vxlan flags */
+	uint8_t reserve1; /**< reserved bits */
+	uint16_t reseve2; /**< reserved bits */
```
[the remark this answers is not in the document]
Comment (bala-manoharan) [reply helper/include/odp/helper/vxlan.h]: Will change the typo.

File: include/odp/api/spec/packet_flags.h [hunk 4]
```diff
@@ -249,6 +249,14 @@ int odp_packet_has_ipsec(odp_packet_t pkt);
  */
 int odp_packet_has_udp(odp_packet_t pkt);
 
+/** Check for VxLAN
```
[the remark this answers is not in the document]
Comment (bala-manoharan) [reply include/odp/api/spec/packet_flags.h]: Okay

File: include/odp/api/spec/packet_flags.h [hunk 5]
```diff
@@ -249,6 +249,14 @@ int odp_packet_has_ipsec(odp_packet_t pkt);
...
+ *
+ * @param pkt Packet handle
+ * @retval non-zero if packet contains VxLAN header
+ * @retval 0 if packet does not contain a VxLAN header
+ */
+int odp_packet_has_vxlan(odp_packet_t pkt);
```
[the remark this answers is not in the document]
Comment (bala-manoharan) [reply include/odp/api/spec/packet_flags.h]: Will migrate to api-next

File: test/validation/api/classification/odp_classification_common.c (hunk 2 above)
Comment (bala-manoharan) [reply test/validation/api/classification/odp_classification_common.c]: Even if we add the capability currently we don't have ODP APIs expose them as packet metadata functions. i.e odp_packet_has_inner_tcp() 

Should we add the capability and defer this to after tiger-moth?

**Comment by muvarov** [comment]
@bala-manoharan PR has to be for api-next patch. @psavol to review api change path.

File: helper/include/odp/helper/vxlan.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+#endif
+
+#include <odp_api.h>
+
+/** @addtogroup odp_header ODP HEADER
+ *  @{
+ */
+
+/** VXLAN header length */
+#define ODPH_VXLANHDR_LEN 8
```
Comment (muvarov) [inline helper/include/odp/helper/vxlan.h]: Please make name more common. ODPH_VXLAN_something everywhere.

File: platform/linux-generic/include/protocols/vxlan.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+#endif
+
+#include <odp_api.h>
+
+/** @addtogroup odp_header ODP HEADER
+ *  @{
+ */
+
+/** VXLAN header length */
+#define _ODP_VXLANHDR_LEN 8
```
Comment (muvarov) [inline platform/linux-generic/include/protocols/vxlan.h]: _ODPH_VXLAN_

File: platform/linux-generic/odp_packet.c [hunk 8]
```diff
@@ -1982,12 +1983,31 @@ static inline void parse_tcp(packet_parser_t *prs,
 	*parseptr += (uint32_t)tcp->hl * 4;
 }
 
+static inline void parse_vxlan(packet_parser_t *prs, packet_parser_t *in_prs,
+			       const uint8_t **parseptr, uint32_t *offset,
+			       uint32_t frame_len, uint32_t seg_len)
+{
+	const _odp_vxlanhdr_t *vxlan = (const _odp_vxlanhdr_t *)*parseptr;
+
+	if (ODPH_VXLAN_BIT(vxlan->flags))
```
Comment (muvarov) [inline platform/linux-generic/odp_packet.c]: do not use helper in platform code. Theoretically that should not even compile. (I hope I will add automatic check for that some day.)

Comment (lumag) [reply platform/linux-generic/odp_packet.c]: @muvarov @psavol 
This issue keeps on risiging again and again. Maybe we can have odp-proto that will be used by odp-helper and by linux-generic (and any other implementation that will need it) but will not be part of ODP API? Code/declaration duplication between linux-generic protocols subdir and helper protocols support is really painful.

**Comment by Bill-Fischofer-Linaro** [comment]
Numerous Travis test failures. Please resolve.

**Comment by Bill-Fischofer-Linaro** [comment]
@bala-manoharan This PR also needs to be rebased to account for the recent forced update to api-next.

**Comment by muvarov** [comment]
@bala-manoharan please rebase PR

**Comment by bala-manoharan** [comment]
To be superseded by design proposal.

**Pull request closed without merging**